File: src/checker/antql/util/qid-unify-util.ts
interface SymbolLike {
  qid?: string
  vtype?: string
  sid?: string
  [key: string]: any
}

/**
 * 统一各语言的qid
 */
class QidUnifyUtil {
  symbol: SymbolLike | undefined

  value: string

  /**
   * 需要传符号值
   * @param symbol
   */
  constructor(symbol?: SymbolLike) {
    this.symbol = symbol
    this.value = symbol?.qid || ''
  }

  /**
   * 统一路径形式，将开头的"/"去掉，并将每一层目录替换成".", 即 /tp/2.func ==> tp.2.func
   */
  removePath(): QidUnifyUtil {
    this.value = this.value?.replace(/^\//, '').replace(/\//g, '.')
    return this
  }

  /**
   * python中找不到import时，会以"syslib_from."开头
   */
  removeSyslibFrom(): QidUnifyUtil {
    if (this.value.startsWith('syslib_from.')) {
      this.value = this.value.replace('syslib_from.', '')
    }
    return this
  }

  /**
   * js-chair框架会将agg替换成Egg.Application，将ctx替换成Egg.Context，替换回来
   */
  removeChair(): QidUnifyUtil {
    this.value = this.value.replace('Egg.Application', 'app')
    this.value = this.value.replace('Egg.Context', 'ctx')
    return this
  }

  /**
   * 去除所有的括号及括号内内容（包括嵌套）——更通用的情况
   */
  removeParentheses(): QidUnifyUtil {
    let result = ''
    let level = 0
    for (const char of this.value) {
      if (char === '(') {
        level++
      } else if (char === ')') {
        if (level > 0) level--
      } else if (level === 0) {
        result += char
      }
    }
    this.value = result
    return this
  }

  /**
   * remove *_scope.<block_>写法，即1.calculate.calculate_scope.<block_18_4_34_51>.process ==> 1.calculate.process
   */
  removeBlock(): QidUnifyUtil {
    if (!this.value.includes('<block')) {
      return this
    }

    // 当符号值类型为symbol时，直接返回sid
    if (this.symbol?.vtype === 'symbol') {
      this.value = this.symbol?.sid || ''
      return this
    }

    const temp = this.value.split('.')
    const result: string[] = []
    for (let i = 0; i < temp.length; i++) {
      const curStr = temp[i]
      const preStr = i > 0 ? temp[i - 1] : 'NaN'
      if (curStr === `${preStr}_scope`) {
        continue
      }
      // 移除掉多余的<block>
      if (curStr.startsWith('<block')) {
        continue
      }
      result.push(curStr)
    }
    this.value = result.join('.')
    return this
  }

  /**
   * 类的实例会表示成*.<instance>.,去掉<instance>
   */
  removeInstance(): QidUnifyUtil {
    this.value = this.value.replace('<instance>', '')
    return this
  }

  /**
   * 统一去掉<global>
   */
  removeGlobal(): QidUnifyUtil {
    this.value = this.value.replace('<global>.', '')
    return this
  }

  /**
   * 获取当前的值
   */
  get(): string {
    return this.value
  }

  /**
   * 静态方法，用于调用上面所有的方法，一步到位统一符号值qid
   * @param symbol
   */
  static unify(symbol?: SymbolLike): string {
    let unifyID = symbol?.qid || ''
    if (symbol?.vtype !== 'primitive' && symbol?.vtype !== 'uninitialized') {
      unifyID = new QidUnifyUtil(symbol)
        .removePath()
        .removeSyslibFrom()
        .removeChair()
        .removeParentheses()
        .removeBlock()
        .removeInstance()
        .removeGlobal()
        .get()
    }
    return unifyID
  }
}

module.exports = QidUnifyUtil

File: yasamcp/src/yasa_mcp/util/qid_unify_util.py
import re


class QidUnifyUtil:
    """
    统一各语言的qid处理工具
    """

    def __init__(self, value):
        self.value = value

    def remove_path(self):
        # 去掉开头"/"，并将所有"/"替换成"."
        if self.value is not None:
            self.value = re.sub(r'^/', '', self.value)
            self.value = self.value.replace('/', '.')
        return self

    def remove_syslib_from(self):
        # 去掉 syslib_from. 前缀
        if self.value and self.value.startswith("syslib_from."):
            self.value = self.value[len("syslib_from."):]
        return self

    def remove_chair(self):
        # Egg.Application -> app, Egg.Context -> ctx
        if self.value:
            self.value = self.value.replace("Egg.Application", "app")
            self.value = self.value.replace("Egg.Context", "ctx")
        return self

    def remove_parentheses(self):
        """
        去除所有括号及括号内内容（包括嵌套），如 foo(bar(baz)) -> foo
        """
        result = ''
        level = 0
        for char in self.value or '':
            if char == '(':
                level += 1
            elif char == ')':
                if level > 0:
                    level -= 1
            elif level == 0:
                result += char
        self.value = result
        return self

    def remove_block(self):
        """
        remove *_scope.<block_>写法，例如 "1.calculate.calculate_scope.<block_18_4_34_51>.process"
        ==> 1.calculate.process
        若 symbol.vtype == "symbol"，直接用 sid
        """
        if not self.value or "<block" not in self.value:
            return self

        temp = self.value.split('.')
        result = []
        for i, cur_str in enumerate(temp):
            pre_str = temp[i - 1] if i > 0 else 'NaN'
            if cur_str == f"{pre_str}_scope":
                continue
            if cur_str.startswith("<block"):
                continue
            result.append(cur_str)
        self.value = ".".join(result)
        return self

    def remove_instance(self):
        # 去掉 <instance>
        if self.value:
            self.value = self.value.replace("<instance>", "")
        return self

    def remove_global(self):
        # 去掉 <global>.
        if self.value:
            self.value = self.value.replace("<global>.", "")
        return self

    def get(self):
        return self.value

    @staticmethod
    def unify(str_value:str):
        unify_id = (QidUnifyUtil(str_value)
                    .remove_path()
                    .remove_syslib_from()
                    .remove_chair()
                    .remove_parentheses()
                    .remove_block()
                    .remove_instance()
                    .remove_global()
                    .get())
        return unify_id


# 用法示例
if __name__ == "__main__":
    # 假设 symbol 是 dict
    value = "com.example.dao.UserDAO<instance>.findById(userId).setName"
    print(QidUnifyUtil.unify(value))  # tp.2.func(app)